GetStream/stream-chat-android
# [4721] Fix infinite retries when a broken link is used

### 🎯 Goal

closes #4271 

### 🛠 Implementation details

Retry only when the connection state changes and the async image painter's state was error due to a timed out request (HTTP code 504)

### 🧪 Testing

_Explain how this change can be tested (or why it can't be tested)_

_Provide a patch below if it is necessary for testing_

<details>

<summary>Patch to be used on the current branch</summary>

```
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt	(revision f4dc6a61ceac1325b509dc144b4757876baee592)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt	(date 1678810409888)
@@ -16,6 +16,7 @@
 
 package io.getstream.chat.android.compose.ui.attachments.content
 
+import android.util.Log
 import androidx.activity.compose.rememberLauncherForActivityResult
 import androidx.compose.foundation.ExperimentalFoundationApi
 import androidx.compose.foundation.Image
@@ -338,16 +339,7 @@
         mutableStateOf(0)
     }
 
-    val data =
-        if (isImage || (isVideo && ChatTheme.videoThumbnailsEnabled)) {
-            when (message.syncStatus) {
-                SyncStatus.COMPLETED ->
-                    attachment.imagePreviewUrl?.applyStreamCdnImageResizingIfEnabled(ChatTheme.streamCdnImageResizing)
-                else -> attachment.upload
-            }
-        } else {
-            null
-        }
+    val data = "notARealUrl"
 
     val painter = rememberStreamImagePainter(
         model = ImageRequest.Builder(LocalContext.current)
@@ -368,6 +360,7 @@
         connectionState = connectionState,
         asyncImagePainterState = painter.state
     ) {
+        Log.d("RetryHash", "Incrementing")
         retryHash++
     }
 
```

</details>


<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt
--- a/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt	(revision 1fa58ed919312846ab84fc880ef293b11db2ee01)
+++ b/stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/content/MediaAttachmentContent.kt	(date 1678810558372)
@@ -16,6 +16,7 @@
 
 package io.getstream.chat.android.compose.ui.attachments.content
 
+import android.util.Log
 import androidx.activity.compose.rememberLauncherForActivityResult
 import androidx.compose.foundation.ExperimentalFoundationApi
 import androidx.compose.foundation.Image
@@ -338,16 +339,7 @@
         mutableStateOf(0)
     }
 
-    val data =
-        if (isImage || (isVideo && ChatTheme.videoThumbnailsEnabled)) {
-            when (message.syncStatus) {
-                SyncStatus.COMPLETED ->
-                    attachment.imagePreviewUrl?.applyStreamCdnImageResizingIfEnabled(ChatTheme.streamCdnImageResizing)
-                else -> attachment.upload
-            }
-        } else {
-            null
-        }
+    val data = "notARealUrl"
 
     val painter = rememberStreamImagePainter(
         model = ImageRequest.Builder(LocalContext.current)
@@ -366,6 +358,7 @@
     if (data != null && connectionState == ConnectionState.CONNECTED &&
         painter.state is AsyncImagePainter.State.Error
     ) {
+        Log.d("RetryHash", "Incrementing")
         retryHash++
     }
 
```

</details>

Test A: Bugfix

1. Apply the patch and run the Compose app (depending on which branch you are running the app from)
2. Enter a channel that has an image in it, or send a message with an image to one
3. Filter the logs for `RetryHash`

Expectation: On `develop`, you should see the logs being constantly filled with `RetryHash`, meaning that the app is continually running requests. On the current branch, you should not see such logs.

Test B: 

1. Without applying the patch, run the Compose app
2. Check that images are loading successfully
3. Turn the network connection off, find an image that has not been loaded and turn the connection on once again. The image should reload automatically

### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

![gif](https://media2.giphy.com/media/Yes4njomZncJi/giphy.gif?cid=ecf05e474g28b6apaik75zvnmwd1a91mb341depehlscalfx&rid=giphy.gif&ct=g)
